runtime/parachains/src/reward_points.rs

```diff
@@ -22,42 +22,71 @@
 //! for the time being, although we will build schemes to do so in the future.
 
 use crate::{session_info, shared};
-use frame_support::traits::ValidatorSet;
-use primitives::v2::ValidatorIndex;
+use frame_support::traits::{Defensive, ValidatorSet};
+use primitives::v2::{SessionIndex, ValidatorIndex};
+use sp_std::collections::btree_set::BTreeSet;
 
 /// The amount of era points given by backing a candidate that is included.
 pub const BACKING_POINTS: u32 = 20;
+/// The amount of era points given by dispute voting on a candidate.
+pub const DISPUTE_STATEMENT_POINTS: u32 = 20;
 
 /// Rewards validators for participating in parachains with era points in pallet-staking.
 pub struct RewardValidatorsWithEraPoints<C>(sp_std::marker::PhantomData<C>);
 
-impl<C> crate::inclusion::RewardValidators for RewardValidatorsWithEraPoints<C>
+impl<C> RewardValidatorsWithEraPoints<C>
 where
-	C: pallet_staking::Config + shared::Config + session_info::Config,
+	C: pallet_staking::Config + session_info::Config,
 	C::ValidatorSet: ValidatorSet<C::AccountId, ValidatorId = C::AccountId>,
 {
-	fn reward_backing(indices: impl IntoIterator<Item = ValidatorIndex>) {
-		// Fetch the validators from the _session_ because sessions are offset from eras
-		// and we are rewarding for behavior in current session.
-		let session_index = shared::Pallet::<C>::session_index();
+	/// Reward validators in session with points, but only if they are in the active set.
+	fn reward_only_active(
+		session_index: SessionIndex,
+		indices: impl IntoIterator<Item = ValidatorIndex>,
+		points: u32,
+	) {
 		let validators = session_info::Pallet::<C>::account_keys(&session_index);
-		let validators = match validators {
+		let validators = match validators
+			.defensive_proof("account_keys are present for dispute_period sessions")
+		{
 			Some(validators) => validators,
-			None => {
-				// Account keys are missing for the current session.
-				// This might happen only for the first session after
-				// `AccountKeys` were introduced via runtime upgrade.
-				return
-			},
+			None => return,
 		};
+		// limit rewards to the active validator set
+		let active_set: BTreeSet<_> = C::ValidatorSet::validators().into_iter().collect();
```

Comment:
Ok: Very stupid question: What is the difference between account_keys and validators?

Also assuming they are different - they seem to contain the same keys (otherwise the contains(v) check would not make any sense) ... but why not use get() on active set directly (before BTreeSetifying it)? The `validators()` call should have correct indices as well - no?

Reply:
For the second part of the question why not:

```rust
let rewards = indices
.into_iter()
.filter_map(|i| validators.get(i.0 as usize).map(|k| (k, points)));
```

with `validators` being:

```
let validators = C::ValidatorSet::validators();
```

Instead of `account_keys`.

Reply:
>  The validators() call should have correct indices as well - no?

No, `ValidatorSet::validators()` is the set of active validators from the session/staking perspective. E.g. 1k on Kusama. 

`account_keys` is parachain validators (200) which can be indexed by `ValidatorIndex` properly.

See also my answer above: https://github.com/paritytech/polkadot/pull/5862#discussion_r945650811